### ISSUE TITLE
[BUG] linear group constraints not upheld

### ISSUE THREAD

**Issue by zoakes** [skfolio/skfolio#82]
**Describe the bug**
I'm looking to create a group-wise equal weight. Anything in group A must be equal weight. 
Maybe this just isn't possible -- but I believe this to be a bug rather than a feature request.

**To Reproduce**
```bash
python3 -m venv env
source env/bin/activate
pip3 install skfolio
```

```python
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from skfolio.optimization import MeanRisk

def rand_returns(n=1000):
    return np.random.normal(loc=0, scale=0.02, size=n)

tdf = pd.DataFrame({
    'a1': rand_returns(), 
    'a2': rand_returns(), 
    'a3': rand_returns(), 
    'b1': rand_returns(), 
    'b2': rand_returns(), 
})

X_train, X_test = train_test_split(tdf, test_size=.33, shuffle=False)

mvo = MeanRisk(
    portfolio_params=dict(name="Force EWGs"),    
    max_weights=1.0,
    groups=[
        # here, I would like all a products to be equal weight. and all b products to be equal weight.
        ['a1','a2','a3', 'b1','b2'], 
    ],
    linear_constraints=[
        # Tried this -- forcing truly equal
        # 'a1 = a2',
        # 'a2 = a3',
        # 'b1 = b2',
    ],
)
mvo.fit(X_train)
mvo_pf = mvo.predict(X_test)
mvo_pf.composition
```

**Expected behavior**
I would expect the composition to be:
a1 x
a2 x 
a3 x

b1 y
b2 y 

i.e. 
```python
mvo_pf.composition.loc['a1'].iloc[0] == mvo_pf.composition.loc['a2'].iloc[0],mvo_pf.composition.loc['a2'].iloc[0] == mvo_pf.composition.loc['a3'].iloc[0],mvo_pf.composition.loc['b1'].iloc[0] == mvo_pf.composition.loc['b2'].iloc[0]

```

**Versions**
Python 3.11.4
skfolio==0.3.1

# full pip freeze:
```
aiodns==3.2.0
aiohttp==3.8.4
aiosignal==1.3.1
anchorpy==0.20.0
anchorpy-core==0.2.0
annotated-types==0.6.0
anyio==4.3.0
appnope==0.1.4
argon2-cffi==23.1.0
argon2-cffi-bindings==21.2.0
arrow==1.3.0
asttokens==2.4.1
async-lru==2.0.4
async-timeout==4.0.2
attrs==23.1.0
Authlib==1.3.0
autopep8==2.1.0
Babel==2.14.0
base58==2.1.1
based58==0.1.1
beautifulsoup4==4.12.3
bitarray==2.9.2
black==23.3.0
bleach==6.0.0
blinker==1.7.0
borsh-construct==0.1.0
build==0.10.0
ccxt==4.3.17
certifi==2023.5.7
cffi==1.16.0
charset-normalizer==3.1.0
ckzg==1.0.1
clarabel==0.9.0
click==8.1.3
comm==0.2.2
construct==2.10.68
construct-typing==0.5.6
contourpy==1.3.0
cryptography==42.0.5
cvxpy==1.5.3
cycler==0.12.1
cytoolz==0.12.3
debugpy==1.8.1
decorator==5.1.1
defusedxml==0.7.1
dexscreener==1.0
dextools-python==0.3.1
docutils==0.20
ecos==2.0.14
eth-account==0.11.2
eth-hash==0.7.0
eth-keyfile==0.8.1
eth-keys==0.5.1
eth-rlp==1.0.1
eth-typing==4.2.3
eth-utils==4.1.1
eth_abi==5.1.0
executing==2.0.1
fastjsonschema==2.19.1
Flask==3.0.3
fonttools==4.53.1
fqdn==1.5.1
frozendict==2.4.4
frozenlist==1.3.3
h11==0.14.0
hexbytes==0.3.1
hmmlearn==0.3.2
html5lib==1.1
httpcore==1.0.5
httpx==0.27.0
idna==3.4
importlib-metadata==6.6.0
iniconfig==2.0.0
ipykernel==6.29.4
ipython==8.24.0
ipywidgets==8.1.2
isoduration==20.11.0
itsdangerous==2.2.0
jaraco.classes==3.2.3
JayDeBeApi==1.2.3
jedi==0.19.1
Jinja2==3.1.3
joblib==1.4.2
JPype1==1.5.0
json5==0.9.25
jsonalias==0.1.1
jsonpointer==2.4
jsonschema==4.21.1
jsonschema-specifications==2023.12.1
jupiter-python-sdk==0.0.2.0
jupyter==1.0.0
jupyter-console==6.6.3
jupyter-events==0.10.0
jupyter-lsp==2.2.5
jupyter_client==8.6.1
jupyter_core==5.7.2
jupyter_server==2.14.0
jupyter_server_terminals==0.5.3
jupyterlab==4.1.8
jupyterlab_pygments==0.3.0
jupyterlab_server==2.27.1
jupyterlab_widgets==3.0.10
keyring==23.13.1
kiwisolver==1.4.7
lru-dict==1.2.0
lxml==5.3.0
markdown-it-py==2.2.0
MarkupSafe==2.1.5
matplotlib==3.9.2
matplotlib-inline==0.1.7
mdurl==0.1.2
mistune==3.0.2
more-itertools==9.1.0
multidict==6.0.4
multitasking==0.0.11
mypy-extensions==1.0.0
nbclient==0.10.0
nbconvert==7.16.3
nbformat==5.10.4
nest-asyncio==1.6.0
notebook==7.1.3
notebook_shim==0.2.4
numpy==1.26.4
oauthlib==3.2.2
osqp==0.6.7.post1
outcome==1.3.0.post0
overrides==7.7.0
packaging==23.1
pandas==2.2.2
pandocfilters==1.5.1
parsimonious==0.10.0
parso==0.8.4
pathspec==0.11.1
peewee==3.17.6
pexpect==4.9.0
pillow==10.4.0
pkginfo==1.9.6
platformdirs==3.5.0
plotly==5.24.0
pluggy==1.0.0
prometheus_client==0.20.0
prompt-toolkit==3.0.43
protobuf==5.26.1
psutil==5.9.8
ptyprocess==0.7.0
pure-eval==0.2.2
pycares==4.4.0
pycodestyle==2.11.1
pycparser==2.22
pycryptodome==3.20.0
pydantic==2.7.1
pydantic_core==2.18.2
Pygments==2.15.1
pyheck==0.1.5
pymssql==2.3.0
pyodbc==5.1.0
pyOpenSSL==24.1.0
pyparsing==3.1.4
pyproject_hooks==1.0.0
PySocks==1.7.1
pytest==7.3.1
pytest-asyncio==0.21.0
pytest-mock==3.10.0
python-dateutil==2.9.0.post0
python-json-logger==2.0.7
pytz==2024.1
pyunormalize==15.1.0
PyYAML==6.0.1
pyzmq==26.0.2
qdldl==0.1.7.post4
qtconsole==5.5.1
QtPy==2.4.1
readme-renderer==37.3
referencing==0.35.0
regex==2024.4.28
requests==2.31.0
requests-oauthlib==2.0.0
requests-toolbelt==1.0.0
rfc3339-validator==0.1.4
rfc3986==2.0.0
rfc3986-validator==0.1.1
rich==13.3.5
rlp==4.0.1
rpds-py==0.18.0
scikit-learn==1.5.1
scipy==1.14.1
scs==3.2.7
selenium==4.19.0
Send2Trash==1.8.3
six==1.16.0
skfolio==0.3.1
sniffio==1.3.1
solana==0.33.0
solders==0.21.0
sortedcontainers==2.4.0
soupsieve==2.5
SQLAlchemy==2.0.31
stack-data==0.6.3
sumtypes==0.1a6
TA-Lib==0.4.32
td-ameritrade-python-api==0.3.5
tda-api==1.6.0
tenacity==9.0.0
terminado==0.18.1
threadpoolctl==3.5.0
tinycss2==1.3.0
toml==0.10.2
toolz==0.11.2
tornado==6.4
traitlets==5.14.3
trio==0.25.0
trio-websocket==0.11.1
twine==4.0.2
types-python-dateutil==2.9.0.20240316
typing_extensions==4.11.0
tzdata==2024.1
uri-template==1.3.0
urllib3==2.0.2
wcwidth==0.2.13
web3==6.18.0
webcolors==1.13
webencodings==0.5.1
websocket-client==1.8.0
websockets==10.4
Werkzeug==3.0.2
whirlpool-essentials==0.0.5
widgetsnbextension==4.0.10
wsproto==1.2.0
yarl==1.9.2
yfinance==0.2.43
zipp==3.15.0
```

**Comment by HugoDelatte** [comment]
Hello, indeed good spot!
The current implementation only supports the strings ">=" or "<=" but not "=" (and it silently ignored it, which is an issue too).

I'm going to fix it.
Also, in the meantime the below solution will work:

```python
mvo = MeanRisk(
    portfolio_params=dict(name="Force EWGs"),
    max_weights=1.0,
    groups=[
        # here, I would like all a products to be equal weight. and all b products to be equal weight.
        ['a1','a2','a3', 'b1','b2'],
    ],
    linear_constraints=[
         'a1 >= a2',
         'a1 <= a2',
         'a2 >= a3',
         'a2 <= a3',
         'b1 >= b2',
         'b1 <= b2',
    ],
)
```

**Comment by zoakes** [comment]
Ah, got it. Thanks!

I’ll give it a shot :)

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by HugoDelatte** [comment]
@zoakes this has been fixed in v0.4.2. Thank you!

**Comment by zoakes** [comment]
Thanks!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
